lib/airbrake-ruby/truncator.rb

```diff
@@ -1,121 +1,96 @@
+    # @return [void]
     def reduce_max_size
       @max_size /= 2
     end
 
     private
 
-    def truncate(val, seen)
-      case val
-      when String
-        truncate_string(val)
-      when Array, Hash, Set
-        truncate_object(val, seen)
-      when Numeric, TrueClass, FalseClass, Symbol, NilClass
-        val
+    def truncate_object(object, seen)
+      case object
+      when Hash then truncate_hash(object, seen)
+      when Array then truncate_array(object, seen)
+      when Set then truncate_set(object, seen)
+      when String then truncate_string(object)
+      when Numeric, TrueClass, FalseClass, Symbol, NilClass then object
       else
-        stringified_val =
-          begin
-            val.to_json
-          rescue *Notice::JSON_EXCEPTIONS
-            val.to_s
-          end
-        truncate_string(stringified_val)
+        truncate_string(stringify_object(object))
       end
     end
 
     def truncate_string(str)
-      str = replace_invalid_characters!(str)
-      return str if str.length <= @max_size
-      str.slice(0, @max_size) + '[Truncated]'.freeze
+      fixed_str = replace_invalid_characters(str)
+      return fixed_str if fixed_str.length <= @max_size
+      (fixed_str[0...@max_size] + '[Truncated]').freeze
     end
 
-    ##
-    # Replaces invalid characters in string with arbitrary encoding.
-    #
-    # @param [String] str The string to replace characters
-    # @return [String] a UTF-8 encoded string
-    # @note This method mutates +str+ unless it's frozen,
-    #   in which case it creates a duplicate
-    # @see https://github.com/flori/json/commit/3e158410e81f94dbbc3da6b7b35f4f64983aa4e3
-    def replace_invalid_characters!(str)
-      encoding = str.encoding
-      utf8_string = (encoding == Encoding::UTF_8 || encoding == Encoding::ASCII)
-      return str if utf8_string && str.valid_encoding?
-
-      str = str.dup if str.frozen?
-      str.encode!(TEMP_ENCODING, ENCODING_OPTIONS) if utf8_string
-      str.encode!('utf-8', ENCODING_OPTIONS)
+    def stringify_object(object)
+      object.to_json
+    rescue *Notice::JSON_EXCEPTIONS
+      object.to_s
     end
 
     def truncate_hash(hash, seen)
+      truncated_hash = {}
       hash.each_with_index do |(key, val), idx|
-        if idx < @max_size
-          hash[key] = truncate(val, seen)
-        else
-          hash.delete(key)
-        end
+        break if idx + 1 > @max_size
+        truncated_hash[key] = truncate(val, seen)
       end
+
+      truncated_hash.freeze
     end
 
     def truncate_array(array, seen)
-      array.slice(0, @max_size).map! { |val| truncate(val, seen) }
+      array[0...@max_size].map { |elem| truncate(elem, seen) }.freeze
     end
 
     def truncate_set(set, seen)
-      set.keep_if.with_index { |_val, idx| idx < @max_size }.map! do |val|
-        truncate(val, seen)
-      end
+      Set.new(truncate_array(set.to_a, seen)).freeze
+    end
```

Comment:
This reads much more cleanly than the original code but I'm a bit worried about potential RAM usage. If this code is running, we must be looking at a Set with more than 10,000 items in it:

* https://github.com/airbrake/airbrake-ruby/blob/master/lib/airbrake-ruby/notice.rb#L80
* https://github.com/airbrake/airbrake-ruby/blob/master/lib/airbrake-ruby/notice.rb#L32

Above, we have the original Set with > 10,000 items, an Array copy with > 10,000 items that we're going to truncate as an Array, creating a 10,000 item Array, which we then convert to a 10,000 item Set.

I'm not very inventive :-) so thinking about how I tackled Hash (and given Set has more in common with Hash, really, than it does Array) we could do something like this:

```ruby
def truncate_set(set, seen)
  truncated_set = Set.new

  set.each do | val |
    truncated_set << truncate(val, seen)
    break if truncated_set.count >= @max_size
  end

  truncated_set.freeze
end
```

It's clearly verbose / clumsy and there may be room for improvement but it doesn't create any intermediate arrays.

Reply:
I like it, it's still simple 👍